### DISCUSSION TITLE
Unreal_v1.0.4 如果继承引擎类的ts类成员变量类型为UActorComponent子类时的想法

### DISCUSSION THREAD

**Discussion by justinlm** [Tencent/puerts#1423] [Ideas]
如果继承引擎类的ts类成员变量类型为UActorComponent子类，将添加组件，而不是仅仅添加一个变量，这会导致业务代码调整

- 定义了组件，就自动在蓝图创建组件，无需在构造函数中通过代码创建，规避了UE的一些多线程加载问题，也更简单些

- 构造函数无法访问Component，建议一些初始化操作放到ReceiveBeginPlay，或者直接在生成的代理蓝图上修改

- 不能通过SetupAttachment对component的层级修改（因为构造函数访问不了，ReceiveBeginPlay又太晚了），需要生成的代理蓝图上手动修改

建议：能不能自动在蓝图创建组件的时候，根据代码里面从上到下或者给个规则，能保证按顺序来生成组件，比如
1. 定义在父类的UActorComponent子类优先创建
2. 定义在自己类里面的UActorComponent子类按照代码里面定义顺序来创建

原因：比如我有这样的使用场景，我在一个组件中定义了很多公用的属性，其他组件都是可以访问的；如果不能保证这个定义了公用属性的组件优先被创建的话，那其他组件去调用这个组件的话，就会出现错误了。

**Comment by chexiongsheng** [comment]
定义UActorComponent后，puerts是调用Editor API往蓝图添加了个Component，和你在蓝图手动添加一个组件是一样的，是一个编辑器行为。
运行时怎么初始化是蓝图的事情。

**Reply by justinlm** [reply]
ok. 明白了
